runtime/lua/vim/lsp/util.lua

```diff
@@ -684,7 +684,20 @@ end
 ---floating window should be created
 function M.focusable_preview(unique_name, fn)
   return M.focusable_float(unique_name, function()
-    return M.open_floating_preview(fn())
+    local contents, filetype, opts = fn()
+    if opts == nil or opts.close.events == nil then
+      -- Provide default autocommand events for closing the preview window
+      opts = opts or {}
+      opts.close_events = {"CursorMoved", "CursorMovedI", "BufHidden"}
+    else
+      -- The "BufLeave" and "WinLeave" events make the preview non-focusable,
+      -- remove them
+      opts.close_events = vim.tbl_filter(
+        function(i) return i ~= 'BufLeave' and i ~= 'WinLeave' end,
+        opts.close_events)
+    end
+
+    return M.open_floating_preview(contents, filetype, opts)
```

Comment:
this logic is pretty confusing to me. Is there any way we can make it less confusing?

For example, we could have `focusable` as a param, and then if that's the case, we could filter out the events you filtered. I do'nt know if that's a good idea, just something I had thought of as reviewing.

Reply:
I've written this comment ~5 times now because I don't really know how to phrase the questions I'm having, so I'll just be blunt: Wouldn't adding another `focusable` check to this already confusing stack of functions (`focusable_preview()`, `open_floating_preview()`, `focusable_float()`, and another ad-hoc closure) make it *even more* confusing?

To be perfectly honest this entire thing feels like a big ball of mud with cross-cutting concerns for no particular gain, and I'd be in favor of simplifying it (IIRC @clason also mentioned that in chat). It's used in only two places in the neovim codebase (`textDocument/hover` uses `focusable_float()`, `textDocument/signatureHelp` uses `focusable_preview()`) and in two corresponding places in completion-nvim. The reason why I haven't attempted simplifying is that I'm not sure how much it is used in plugins and whatnot. It *is* exposed in `lsp.util` now, after all. Then again, this is nightly, so… I don't know.

Any input on how I should go about this is appreciated. :-/

Reply:
I've removed this logic, instead I just schedule the autocmds on the libuv loop so that they get set *after* the new window is focused.